ext/digest/crc/crc_ext.c

```diff
@@ -0,0 +1,431 @@
+	int n = NUM2INT(v);
+
+	if (n < 1 || n > 64) {
+		rb_raise(rb_eNotImpError,
+		         "%" PRIsVALUE " is incompleted as CRC (WIDTH must be 1..64)",
+		         model_class);
+	}
+
+	return n;
+}
+
+static uint64_t ensure_generator_polynomial(VALUE model_class)
+{
+	VALUE v = rb_const_get(model_class, rb_intern("POLYNOMIAL"));
+	uint64_t n = NUM2ULL(v);
+
+	if ((n & 1) == 0) {
+		rb_raise(rb_eNotImpError,
+		         "%" PRIsVALUE " is incompleted as CRC (POLYNOMIAL must be odd number)",
+		         model_class);
+	}
+
+	return n;
+}
+
+static int ensure_reflect_input(VALUE model_class)
+{
+	VALUE v = rb_const_get(model_class, rb_intern("REFLECT_INPUT"));
+
+	if (v != Qtrue && v != Qfalse) {
+		rb_raise(rb_eNotImpError,
+		         "%" PRIsVALUE " is incompleted as CRC (REFLECT_INPUT must be true or false)",
+		         model_class);
+	}
+
+	return RTEST(v);
+}
+
+static const struct crc_model *get_crc_model(VALUE self)
+{
+	ID id_model = rb_intern("digest-crc-model");
+	VALUE model_class = rb_class_of(self);
+	VALUE model_value = rb_ivar_get(model_class, id_model);
+	struct crc_model *model;
+
+	if (RB_TYPE_P(model_value, RUBY_T_DATA) && RTYPEDDATA_P(model_value)) {
+		const struct rb_data_type_struct *type = RTYPEDDATA_TYPE(model_value);
+		if (type == &crc_model_type ||
+		    (type && strcmp(type->wrap_struct_name, "digest-crc:model-const") == 0)) {
+			model = (struct crc_model *)RTYPEDDATA_DATA(model_value);
+			if (model) {
+				return model;
+			}
+		}
+	}
+
+	ruby_digest_crc_ensure_ractor_main("only the main Ractor can initialize tables");
+
+	int bitwidth = ensure_bitwidth(model_class);
+	uint64_t polynomial = ensure_generator_polynomial(model_class);
+	int refin = ensure_reflect_input(model_class);
+
+	rb_ivar_set(model_class, id_model, Qfalse); // ensured not freezing, and reserved the iv
+	model_value = rb_data_typed_object_zalloc(rb_cObject, sizeof(struct crc_model), &crc_model_type);
+	model = (struct crc_model *)RTYPEDDATA_DATA(model_value);
+	model->bitwidth = bitwidth;
+	model->reflect_input = refin;
+	model->polynomial = polynomial;
+	model->table_heap = crc_alloc_heap;
+	initialize_crc_table(model);
+	rb_obj_freeze(model_value);
+	rb_ivar_set(model_class, id_model, model_value);
+
+	return model;
+}
+
+static VALUE digest_crc_update(VALUE self, VALUE str)
+{
+	ID id_iv_crc = rb_intern("@crc");
+	VALUE crc_value = rb_ivar_get(self, id_iv_crc);
+	const struct crc_model *model = get_crc_model(self);
+	const char *ptr = StringValuePtr(str);
+	size_t len = RSTRING_LEN(str);
+
+	crc_value = ULL2NUM(update_crc(model, NUM2ULL(crc_value), ptr, len));
+	rb_ivar_set(self, id_iv_crc, crc_value);
+
+	return self;
+}
+
+void Init_crc_ext(void)
+{
+	#ifdef HAVE_RB_EXT_RACTOR_SAFE
+		rb_ext_ractor_safe(true);
+	#endif
+
+	VALUE mDigest = rb_const_get(rb_cObject, rb_intern("Digest"));
+	VALUE cCRC = rb_const_get(mDigest, rb_intern("CRC"));
+
+	rb_define_method(cCRC, "update", digest_crc_update, 1);
+}
```

Comment:
I understand the strategy.
Then, I will have a dedicated code for each of them.

Reply:
I think always auto-generating the code would be more similar to how most C/C++ project's use `./configure`.

Alternatively, we could generate both the 8bit and 16bit slicing implementations in separate files or directories and commit them to git, and have `extconf.rb` merely select which file to `#include`. If the generated 8bit and 16bit slicing code will never change and doesn't rely on any other dynamic user configuration, then we could just generate both implementations once.

Reply:
Also, both tables could be generated in the same file but guarded with a `#ifdef`/`#else` to select which table to use based on some `-D` macro.

Reply:
Regarding the bloat of the GEM package, is it acceptable?

I actually added a Slicing-by-16 table file and the GEM package was 521 KiB.

```console
% git add 'ext/digest/*/*_table.h' && git ci -am 'temp' && rake build`
...

% ll pkg
total 1
-rw-r--r--  1 dearblue  dearblue  533504  7月 26 21:33 digest-crc-0.6.5.gem
```

:smiley_cat: And the "N" in "Slicing-by-N" is probably derived from "Byte".

Reply:
512 KiB for all generated tables doesn't seem that bad. If the `.gem` package became larger than a few Mbs, then maybe we should auto-generate the 16bit sliced tables to reduce total package size.

I mainly want to ensure all code is there in the git repo, in case anyone ever wants to audit the code or run it through some kind of static analyzer.